control-plane/csi-controller/src/client.rs

```diff
@@ -0,0 +1,373 @@
+                "Failed to obtain body from HTTP response while getting {}, error = {}",
+                urn, e,
+            ))
+        })?;
+
+        serde_json::from_slice::<R>(&body).map_err(|e| {
+            ApiClientError::InvalidResponseError(format!(
+                "Failed to deserialize object {}, error = {}",
+                std::any::type_name::<R>(),
+                e
+            ))
+        })
+    }
+
+    // Get one resource instance.
+    async fn do_get(&self, urn: &UrnType<'_>) -> Result<Response, Error> {
+        let u = format!("{}/{}", self.base_url, urn);
+        let uri = Url::parse(&u).unwrap();
+
+        self.rest_client.get(uri).send().await
+    }
+
+    // Perform resource deletion, optionally idempotent.
+    async fn do_delete(&self, urn: &UrnType<'_>, idempotent: bool) -> Result<(), ApiClientError> {
+        let u = format!("{}/{}", self.base_url, urn);
+        let uri = Url::parse(&u).unwrap();
+
+        let response = self.rest_client.delete(uri).send().await.map_err(|e| {
+            ApiClientError::ServerCommunicationError(format!(
+                "DELETE {} request failed, error={}",
+                u, e
+            ))
+        })?;
+
+        // Check HTTP status code, handle DELETE idempotency transparently.
+        let res = match response.status() {
+            StatusCode::OK => Ok(()),
+            // Handle idempotency as requested by the caller.
+            StatusCode::NOT_FOUND | StatusCode::NO_CONTENT | StatusCode::PRECONDITION_FAILED => {
+                if idempotent {
+                    Ok(())
+                } else {
+                    let (rtype, rname) = urn.classify();
+                    return Err(ApiClientError::ResourceNotExists(format!(
+                        "{} {} not found",
+                        rtype, rname
+                    )));
+                }
+            }
+            code => Err(ApiClientError::GenericOperationError(format!(
+                "DELETE {} failed, HTTP status code = {}",
+                u, code
+            ))),
+        };
+        debug!("Resource {} successfully deleted", u);
+        res
+    }
+
+    async fn do_put<I, O>(&self, urn: &UrnType<'_>, object: I) -> Result<O, ApiClientError>
+    where
+        I: Serialize + Sized,
+        for<'a> O: Deserialize<'a>,
+    {
+        let u = format!("{}/{}", self.base_url, urn);
+        let uri = Url::parse(&u).unwrap();
+
+        let response = self
+            .rest_client
+            .put(uri)
+            .json(&object)
+            .send()
+            .await
+            .map_err(|e| {
+                ApiClientError::ServerCommunicationError(format!(
+                    "PUT {} request failed, error={}",
+                    u, e
+                ))
+            })?;
+
+        // Check HTTP status of the operation.
+        match response.status() {
+            StatusCode::OK => {}
+            StatusCode::UNPROCESSABLE_ENTITY => {
```

Comment:
Did you end up creating a ticket to fix these error codes? If so could you please add a todo: CAS-XXX where it applies so that we know what to change when we fix the ticket?

Reply:
Do we need to fix the errors returned by Control Plane ?

Reply:
I created CAS-1124 as a placeholder and a TODO.